src/query/functions/temporal/base.go

```diff
@@ -0,0 +1,505 @@
+
+	// Process right side of the range
+	rightBlks, emptyRightBlocks, err := c.processRight(bounds, maxBlocks, rightRangeStart)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(rightBlks); i++ {
+		lStart := maxBlocks - i
+		if lStart > len(leftBlks) {
+			continue
+		}
+
+		deps := leftBlks[len(leftBlks)-lStart:]
+		deps = append(deps, rightBlks[:i]...)
+		processRequests = append(processRequests, processRequest{blk: rightBlks[i], deps: deps, bounds: bounds.Next(i + 1)})
+	}
+
+	// If either the left range or right range wasn't fully processed then cache the current block
+	if emptyLeftBlocks || emptyRightBlocks {
+		if err := c.cache.add(bounds.Start, b); err != nil {
+			return err
+		}
+	}
+
+	return c.processCompletedBlocks(processRequests, queryStartBounds, queryEndBounds, maxBlocks)
+}
+
+// processCurrent processes the current block. For the current block, figure out whether we have enough previous blocks which can help process it
+func (c *baseNode) processCurrent(bounds block.Bounds, maxBlocks int, leftRangeStart block.Bounds) ([]block.Block, bool, error) {
+	numBlocks := bounds.Blocks(leftRangeStart.Start)
+	leftBlks, err := c.cache.multiGet(leftRangeStart, numBlocks, true)
+	if err != nil {
+		return nil, false, err
+	}
+	return leftBlks, len(leftBlks) != numBlocks, nil
+}
+
+// processRight processes blocks after current block. This is done by fetching all contiguous right blocks until the right range
+func (c *baseNode) processRight(bounds block.Bounds, maxBlocks int, rightRangeStart block.Bounds) ([]block.Block, bool, error) {
+	numBlocks := rightRangeStart.Blocks(bounds.Start)
+	rightBlks, err := c.cache.multiGet(bounds.Next(1), numBlocks, false)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return rightBlks, len(rightBlks) != maxBlocks, nil
+}
+
+// processCompletedBlocks processes all blocks for which all dependent blocks are present
+func (c *baseNode) processCompletedBlocks(processRequests []processRequest, queryStartBounds, queryEndBounds block.Bounds, maxBlocks int) error {
+	processedKeys := make([]time.Time, len(processRequests))
+	for i, req := range processRequests {
+		if err := c.processSingleRequest(req); err != nil {
+			return err
+		}
+
+		processedKeys[i] = req.bounds.Start
+	}
+
+	// Mark all blocks as processed
+	c.cache.markProcessed(processedKeys)
+	// Sweep to free blocks from cache with no dependencies
+	c.sweep(c.cache.processed(), queryStartBounds, queryEndBounds, maxBlocks)
+	return nil
+}
+
+func (c *baseNode) processSingleRequest(request processRequest) error {
+	aggDuration := c.op.duration
+	seriesIter, err := request.blk.SeriesIter()
+	if err != nil {
+		return err
+	}
+
+	depIters := make([]block.SeriesIter, len(request.deps))
+	for i, blk := range request.deps {
+		iter, err := blk.SeriesIter()
+		if err != nil {
+			return err
+		}
+		depIters[i] = iter
+	}
+
+	bounds := seriesIter.Meta().Bounds
+	steps := int((aggDuration + bounds.Duration) / bounds.StepSize)
```

Comment:
Add stepSize==0 sanity check; also can define this closer to where it's needed

Reply:
should be handled by sanity check on bounds in Process()

Reply:
need in the next line